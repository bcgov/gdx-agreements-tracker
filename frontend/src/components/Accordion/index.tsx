import * as React from "react";
import { Accordion as MUIAccordion } from "@mui/material/";
import AccordionSummary from "@mui/material/AccordionSummary";
import AccordionDetails from "@mui/material/AccordionDetails";
import Typography from "@mui/material/Typography";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import { styled } from "@mui/material";
import bcgovTheme from "../../bcgovTheme";

export const Accordion = ({
  children,
  sectionTitle,
}: {
  children: JSX.Element | JSX.Element[];
  sectionTitle: string;
}) => {
  const StyledAccordionSummary = styled(AccordionSummary)({
    backgroundColor: "#f3f3f3",
  });

  const StyledTypography = styled(Typography)({
    color: bcgovTheme.palette.primary.main,
    fontWeight: "bold",
  });

  return (
    <MUIAccordion defaultExpanded>
      <StyledAccordionSummary
        expandIcon={<ExpandMoreIcon />}
        aria-controls="panel-content"
        aria-label="button"
      >
        <StyledTypography>{sectionTitle}</StyledTypography>
      </StyledAccordionSummary>
      <AccordionDetails>{children}</AccordionDetails>
    </MUIAccordion>
  );
};
