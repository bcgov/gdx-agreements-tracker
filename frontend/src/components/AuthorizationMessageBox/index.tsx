import { FormHelperText } from "@mui/material";

export const AuthorizationMessageBox = ({ message }: { message: string }) => {
  return <FormHelperText>{message}</FormHelperText>;
};
