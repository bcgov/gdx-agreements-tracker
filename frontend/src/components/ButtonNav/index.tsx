import React from "react";
import { Link, useMatch, useResolvedPath } from "react-router-dom";
import type { LinkProps } from "react-router-dom";

/**
 * Returns react router link with active class if matches route.
 *
 * @param {LinkProps} root0 Link properties.
 * @param {ReactNode | undefined} root0.children Child nodes.
 * @param {To} root0.to Destination.
 * @returns {React.ReactNode}
 */
export const ButtonNav = ({ children, to, ...props }: LinkProps) => {
  const resolved = useResolvedPath(to);
  const match = useMatch({ path: resolved.pathname, end: true });
  if (match) {
    props.className = `${props.className ?? ""} active`;
  }
  return (
    <Link to={to} {...props}>
      {children}
    </Link>
  );
};

export default ButtonNav;
