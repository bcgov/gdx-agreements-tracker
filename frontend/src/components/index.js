export { ButtonNav } from "./ButtonNav";
export { Debug } from "./Debug";
export { Loader } from "./Loader";
export { Main } from "./Layout/Main";
export { PageLink } from "./PageLink";
export { RowEditDeleteChip } from "./RowEditDeleteChip";
export { Sidebar } from "./Sidebar";
export { DetailsList } from "./DetailsList";
export { Table } from "./Table";
