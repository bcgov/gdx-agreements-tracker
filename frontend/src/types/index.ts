import { GridRowsProp, GridColDef } from "@mui/x-data-grid";

// Data Structures
export interface IUser {
  created_at: string;
  email: string;
  id: number;
  name: string;
  updated_at: string;
  username: string;
  roles?: unknown;
}

// Hooks
export interface IUseDrawer {
  drawerOpen: boolean;
  handleDrawerToggle: (event: React.MouseEvent<HTMLElement>) => void;
}

// Components
export interface IPageHeader extends IUseDrawer {}

export interface ISidebar extends IUseDrawer {}

export interface ITable {
  rows: GridRowsProp;
  columns: GridColDef[];
  loading: boolean;
}

// Tables
export interface IColumn {
  id: number;
  field: string;
  headerName: string;
  flex: number;
}

export interface ITableData {
  data: Array<Object>;
}

//Declartion that adds custom types to material ui predefined interfaces
declare module "@mui/material/styles" {
  interface Theme {
    customSettings: {
      BCGovAccentLine: string;
      topBarHeight: string;
      drawerWidth: number;
    };
  }
  // allow configuration using `createTheme`
  interface ThemeOptions {
    customSettings?: Theme["customSettings"];
  }
}

// Project layout types
export interface IFormLayout {
  children: JSX.Element;
}

//Picker Types
export interface IPickerLookupData {
  id: number;
  name: string;
  title: string;
  description: string;
  definition: { dropDownValues: { data: Array<Object> } };
}

//picker options types
export interface IPickerProps {
  handleChange: Function;
  formikValues: { [key: string]: unknown };
  setFieldValue: Function;
  pickerData: {
    associated_table: string;
    definition: { value: string; label: string }[];
    description: string;
    id: number;
    name: string;
    title: string;
  };
}

export interface IPickerTableData {
  data: [
    {
      associated_table: string;
      definition: [
        {
          label: string;
          value: string;
        }
      ];
      description: string;
      id: number;
      name: string;
      title: string;
    }
  ];
}

export interface IProjectLayout {
  children: JSX.Element;
}

//Picker Types
export interface IPickerLookupData {
  id: number;
  name: string;
  title: string;
  description: string;
  definition: { dropDownValues: { data: Array<Object> } };
}

//ChipNav Types
export interface IChipNav {
  navLinks: Array<{ key: number; name: string; url: string }>;
}
