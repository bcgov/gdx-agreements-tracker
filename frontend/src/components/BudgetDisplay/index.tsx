import { Box, Grid } from "@mui/material";
import { GDXList } from "components/GDXList";
import { Loader } from "components/Loader";
import { useAxios } from "hooks/useAxios";
import React from "react";
import { useQuery } from "react-query";

export const BudgetDisplay = ({ apiUrl }: { apiUrl: string }) => {
  const { axiosAll } = useAxios();

  const getBudgetData = async () => {
    const budget = await axiosAll()
      .get(apiUrl)
      .then((budgetData: any) => {
        return budgetData;
      });
    return budget.data.data;
  };

  // Queries
  const { data, isLoading } = useQuery("contract budget", getBudgetData, {
    // todo: When there is an edit and view form built, reassess these options.
    refetchOnWindowFocus: false,
    retryOnMount: false,
    refetchOnReconnect: false,
    retry: false,
    staleTime: Infinity,
  });

  return (
    <Grid container spacing={2}>
      {!data ? (
        <Loader />
      ) : (
        data.map((budget: any) => {
          return <GDXList data={budget} title={`Fiscal ${budget.fiscal_year}`} />;
        })
      )}
    </Grid>
  );
};