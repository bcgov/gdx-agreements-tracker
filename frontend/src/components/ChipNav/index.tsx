import * as React from "react";
import { styled } from "@mui/material/styles";
import Chip from "@mui/material/Chip";
import { IChipNav } from "../../types";
import { Box, Paper } from "@mui/material";
import { NavLink } from "react-router-dom";
import bcgovTheme from "../../bcgovTheme";

export const ChipNav = ({
  navLinks,
  navLinksRight = [],
}: {
  navLinks: IChipNav[];
  navLinksRight?: IChipNav[];
}) => {
  const StyledListItem = styled("li")(({ theme }) => ({
    margin: theme.spacing(0.5),
  }));

  const StyledNavLink = styled(NavLink)(() => ({
    textDecoration: "none",
  }));

  /**
   * Renders a single ChipNav link.
   *
   * @param   {IChipNav}    link The ChipNav link to render.
   * @returns {JSX.Element}
   */
  const renderNavLink = (link: IChipNav): JSX.Element => {
    return (
      <StyledListItem key={link.key}>
        <StyledNavLink to={link.url} end>
          {({ isActive }) => (
            <Chip
              sx={{
                backgroundColor: isActive ? "#D3D3D3" : bcgovTheme.palette.primary.main,
                color: isActive ? "#000" : "#FFF",
              }}
              label={link.name}
            />
          )}
        </StyledNavLink>
      </StyledListItem>
    );
  };

  return (
    <Paper
      elevation={3}
      sx={{
        display: "flex",
        justifyContent: "space-between",
        flexWrap: "wrap",
        p: 0.5,
        mb: 2,
        mt: 0,
      }}
    >
      <Box sx={{ display: "flex", listStyle: "none" }}>{navLinks.map(renderNavLink)}</Box>
      {navLinksRight.length > 0 && (
        <Box sx={{ display: "flex", listStyle: "none" }}>{navLinksRight.map(renderNavLink)}</Box>
      )}
    </Paper>
  );
};
