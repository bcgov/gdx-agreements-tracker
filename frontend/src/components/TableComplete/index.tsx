import React, { useEffect, useState } from "react";
import { Box, Button, Typography } from "@mui/material";
import { Renderer } from "components/Renderer";
import { Table } from "components/Table";
import { useFormControls } from "hooks/useFormControls";
import { useFormatTableData } from "hooks";
import { useFormSubmit } from "hooks/useFormSubmit";
import { GDXModal } from "components/GDXModal";
import { useQuery, UseQueryResult } from "react-query";
import { FormikValues } from "formik";
import { ReadForm } from "components/ReadForm";
import { CreateForm } from "components/CreateForm";
import { EditForm } from "components/EditForm";
import { useAxios } from "hooks/useAxios";
import { IEditField, IInitialValues, IReturnValue } from "types";
import { GridEventListener, GridEvents } from "@mui/x-data-grid";

export const TableComplete = ({
  itemName,
  tableName,
  url,
  createFormInitialValues,
  readFields,
  editFields,
  totalColumns,
  roles,
  getSelectedRow,
  columnWidths,
}: {
  itemName: string;
  tableName: string;
  url: {
    getAll: string;
    getOne: string;
    updateOne: string;
    addOne: string;
    deleteOne?: string;
  };
  createFormInitialValues: IInitialValues;
  readFields: Function;
  editFields: IEditField[];
  totalColumns?: string[];
  roles: {
    get: string;
    update: string;
    add: string;
    delete: string;
  };
  getSelectedRow?: Function;
  columnWidths?: { [key: string]: number };
}) => {
  const {
    handleEditMode,
    handleOpen,
    handleClose,
    handleCurrentRowData,
    handleFormType,
    formType,
    open,
    editMode,
    currentRowData,
  } = useFormControls();


  const { handlePost, handleUpdate, Notification } = useFormSubmit();

  const [userCapabilities, setUserCapabilities] = useState<string[]>([]);

  const { axiosAll } = useAxios();

  /**
   * returns destructured props from the useFormatTableData hook.
   *
   * @param   {string}   tableName   - The name of the table that you are wanting data from.
   * @param   {string}   apiEndPoint - The enpoint as which the API query will use for it's call.
   * @param   {Function} handleClick - Function passed to the "view" button of the Table component.
   * @returns {object}               {data, isLoading}  - "data" contains the columns and rows of data for your table.  isLoading is a boolean prop that changes to true if quering data and false if it has received the data.
   */

  const { data, isLoading } = useFormatTableData({
    tableName: tableName,
    apiEndPoint: url.getAll,
    columnWidths: columnWidths,
    handleClick: handleOpen,
  });

  

  useEffect(() => {
    setUserCapabilities(data?.user?.capabilities);
  }, [data]);

  /**
   * getApiData is the fetch function for react query to leverage.
   *
   * @returns {object} An object that contains the data from the table it's querying.
   */
  const getApiData = async () => {
    const apiUrl = getApiUrl(url.getOne, currentRowData?.id);
    if (currentRowData?.id && apiUrl) {
      const apiData = await axiosAll().get(apiUrl);
      return apiData.data.data;
    }
  };

  /**
   * Get the url for a getOne item.
   *
   * @param   {string}           url The raw url, that might need to be converted to replace id with current id.
   * @param   {number|undefined} id  The current id, used to create the url.
   * @returns {string}
   */
  const getApiUrl = (url: string, id: number | undefined) => {
    if (undefined !== id) {
      return url.replace(/{id}/g, id.toString());
    }
    return "";
  };

  const handleDeleteClick = () => {
    if (!url.deleteOne) {
      return false;
    }
    handleDelete({
      apiUrl: getApiUrl(url.deleteOne as string, currentRowData.id),
      queryKeys: [url.getAll],
      successMessage: "Deleted successfully.",
      errorMessage: "There was an issue deleting the item.",
    });
  };

  /**
   * returns destructured props from the useFormatTableData hook.
   *
   * @param   {string}   tableName   - The name of the table that you are wanting data from.
   * @param   {string}   apiEndPoint - The enpoint as which the API query will use for it's call.
   * @param   {Function} handleClick - Function passed to the "view" button of the Table component.
   * @returns {object}               {data, isLoading}  - "data" contains the columns and rows of data for your table.  isLoading is a boolean prop that changes to true if quering data and false if it has received the data.
   */
  const { data, isLoading } = useFormatTableData({
    tableName: tableName,
    apiEndPoint: url.getAll,
    columnWidths: columnWidths,
    handleClick: handleOpen,
  });

  useEffect(() => {
    setUserCapabilities(data?.user?.capabilities);
  }, [data]);

  /**
   * used for the react query.
   *
   * @param   {string}         queryKey     - This is the queryKey.  The queryKey acts as a cache identifier for the UseQueryResult.
   * @param   {Function}       getAmendment - The enpoint as which the API query will use for it's call.
   * @returns {UseQueryResult}              - The result of react query which contains things such as the data.
   */
  // Queries
  export const reactQuery: UseQueryResult<FormikValues> = useQuery(
    getApiUrl(url.getOne, currentRowData?.id),
    getApiData,
    {
      refetchOnWindowFocus: false,
      retryOnMount: false,
      refetchOnReconnect: false,
      retry: false,
      staleTime: Infinity,
    }
  );

  /**
   * Handle row click event from DataGrid.
   *
   * @param {any} row The row that was clicked.
   */
  const handleRowClick: GridEventListener<GridEvents.rowClick> = (row) => {
    handleCurrentRowData(row);
    if (getSelectedRow) {
      getSelectedRow(row);
    }
  };

  return (
    <>
      {hasRole(roles.get) && (
        <Renderer
          isLoading={isLoading}
          component={
            <>
              <Typography variant="h5" component="h2">
                {itemName}
              </Typography>
              <Table
                columns={data?.columns}
                rows={data?.rows}
                totalColumns={totalColumns ?? []}
                allowEdit={hasRole(roles.add)}
                loading={isLoading}
                onRowClick={handleRowClick}
              />
              <Box m={1} display="flex" justifyContent="flex-end" alignItems="flex-end">
                {hasRole(roles.add) && (
                  <Button
                    onClick={() => {
                      handleOpen();
                      handleEditMode(true);
                      handleFormType("new");
                    }}
                    variant="contained"
                  >{`Add New ${itemName}`}</Button>
                )}
              </Box>
            </>
          }
        />
      )}
      <GDXModal
        open={open}
        handleClose={handleClose}
        modalTitle={"new" === formType ? `New ${itemName}` : `${itemName} ${reactQuery?.data?.id}`}
        handleEditMode={handleEditMode}
        editMode={editMode}
        allowEdit={hasRole(roles.update)}
        allowDelete={hasRole(roles.delete) && undefined !== url.deleteOne && "new" !== formType}
        handleDelete={handleDeleteClick}
        handleFormType={handleFormType}
      >
        <>
          {!editMode ? (
            <ReadForm fields={readFields(reactQuery)} />
          ) : (
            <>
              {"new" === formType ? (
                <CreateForm
                  initialValues={createFormInitialValues}
                  onSubmit={async (values: { [key: string]: IReturnValue }) => {
                    return handlePost({
                      formValues: values,
                      apiUrl: url.addOne,
                      handleEditMode: handleEditMode,
                      queryKeys: [url.getAll],
                      successMessage: `Created successfully.`,
                      errorMessage: `There was an issue creating your item.`,
                      handleClose: handleClose,
                    });
                  }}
                  editFields={editFields}
                />
              ) : (
                <EditForm
                  initialValues={reactQuery?.data as FormikValues}
                  onSubmit={async (values) => {
                    return handleUpdate({
                      changedValues: values,
                      currentRowData: reactQuery?.data,
                      apiUrl: getApiUrl(url.updateOne, reactQuery?.data?.id),
                      handleEditMode: handleEditMode,
                      queryKeys: [url.getAll, getApiUrl(url.getOne, reactQuery?.data?.id)],
                      successMessage: `Changes saved successfully.`,
                      errorMessage: `There was an issue saving.`,
                    });
                  }}
                  editFields={editFields}
                />
              )}
            </>
          )}
        </>
      </GDXModal>
      <Notification />
    </>
  );
};
