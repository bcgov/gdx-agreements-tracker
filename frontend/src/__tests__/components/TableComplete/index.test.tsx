import React from "react";
import { mount } from "enzyme";
import { TableComplete } from "components/TableComplete";
import { IEditFields } from "types";
import { FormikValues } from "formik";
import { UseQueryResult } from "react-query";
import { createMemoryHistory } from "history";
import { Router } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "react-query";

const readFields = (ministriesQuery: UseQueryResult<FormikValues>) => {
  return [
    { width: "half", title: "Ministry", value: ministriesQuery?.data?.ministry_name },
    { width: "half", title: "Abbr", value: ministriesQuery?.data?.ministry_short_name },
    { width: "half", title: "Is Active", value: ministriesQuery?.data?.is_active ? "Yes" : "No" },
  ];
};

const editFields: IEditFields[] = [
  {
    fieldName: "ministry_name",
    fieldType: "singleText",
    fieldLabel: "Name",
    width: "half",
  },
  {
    fieldName: "ministry_short_name",
    fieldType: "singleText",
    fieldLabel: "Abbreviation",
    width: "half",
  },
  {
    fieldName: "is_active",
    fieldType: "checkbox",
    fieldLabel: "Is Active",
    width: "half",
  },
];

const createFormInitialValues = {
  ministry_name: "",
  ministry_short_name: "",
  is_active: false,
};

const roles = {
  get: "admin_form_read_all",
  add: "admin_form_add_one",
  update: "admin_form_update_one",
  delete: "admin_form_delete_one",
};

const url = {
  getAll: `/ministries`,
  getOne: `/ministries/{id}`,
  updateOne: `/ministries/{id}`,
  addOne: `/ministries`,
};

const queryClient = new QueryClient();

//Mock keycloak.
jest.mock("@react-keycloak/web", () => ({
  useKeycloak: () => ({ initialized: true, keycloak: { authenticated: true } }),
}));

describe("Tests different renders of the <Renderer /> component", () => {
  const history = createMemoryHistory();
  //This function allows you to render the component with different isLoading props to test different outcomes
  const wrapper = mount(
    <QueryClientProvider client={queryClient}>
      <Router location={history.location} navigator={history}>
        <TableComplete
          itemName="Ministry"
          tableName="ministry"
          url={url}
          createFormInitialValues={createFormInitialValues}
          readFields={readFields}
          editFields={editFields}
          roles={roles}
        />
      </Router>
    </QueryClientProvider>
  );

  it("test 1", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
