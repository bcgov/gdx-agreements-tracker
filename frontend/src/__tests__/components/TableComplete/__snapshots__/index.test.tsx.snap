// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests different renders of the <Renderer /> component test 1 1`] = `""`;
